File: examples/ts/search_indexer.ts
/**
 * Search indexer - build inverted index from Japanese text
 *
 * Demonstrates: tag generation with POS filtering, batch processing,
 * and building a search index in the browser/Node.js.
 *
 * Run: npx tsx examples/ts/search_indexer.ts
 */
import { Suzume } from '../../bindings/wasm/dist/index.js';

const suzume = await Suzume.create();

interface Document {
  id: number;
  text: string;
}

const docs: Document[] = [
  { id: 1, text: '東京都渋谷区で新しいカフェがオープンしました' },
  { id: 2, text: '渋谷のカフェで美味しいコーヒーを飲みました' },
  { id: 3, text: '新宿駅の近くにある図書館で本を読んでいます' },
];

// Build inverted index: tag -> Set<doc_id>
const index = new Map<string, Set<number>>();

for (const doc of docs) {
  const tags = suzume.generateTags(doc.text, {
    posFilter: ['noun', 'verb'],
    excludeBasic: true,
    useLemma: true,
    minLength: 2,
  });
  if (tags.length === 0) {
    throw new Error(`tag generation returned no tags for document ${doc.id}`);
  }

  for (const tag of tags) {
    const docIds = index.get(tag.tag) ?? new Set<number>();
    docIds.add(doc.id);
    index.set(tag.tag, docIds);
  }
}

// Print index
console.log('Inverted index:');
for (const [tag, docIds] of index) {
  console.log(`  ${tag} -> [${[...docIds].join(', ')}]`);
}

// Search
function search(query: string): number[] {
  const queryTags = suzume.generateTags(query, {
    posFilter: ['noun', 'verb'],
    useLemma: true,
  });

  const scores = new Map<number, number>();
  for (const tag of queryTags) {
    const docIds = index.get(tag.tag);
    if (docIds) {
      for (const id of docIds) {
        scores.set(id, (scores.get(id) ?? 0) + 1);
      }
    }
  }

  return [...scores.entries()].sort((a, b) => b[1] - a[1]).map(([id]) => id);
}

console.log('\nSearch "渋谷 カフェ":', search('渋谷 カフェ'));
console.log('Search "図書館 読む":', search('図書館 読む'));
if (index.size === 0) {
  throw new Error('search index is empty');
}

File: examples/cpp/search_indexer.cpp
/**
 * Search indexer - build inverted index from Japanese text
 *
 * Demonstrates: tag generation, POS filtering, deduplication,
 * and batch processing of multiple documents.
 */
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

#include "suzume/suzume.hpp"

struct Document {
  int id;
  std::string text;
};

int main() {
  suzume::Tokenizer analyzer;
  if (!analyzer) {
    std::cerr << "failed to create tokenizer: " << suzume::Tokenizer::lastError() << "\n";
    return 1;
  }

  // Sample documents
  std::vector<Document> docs = {
      {1, "東京都渋谷区で新しいカフェがオープンしました"},
      {2, "渋谷のカフェで美味しいコーヒーを飲みました"},
      {3, "新宿駅の近くにある図書館で本を読んでいます"},
  };

  // Build inverted index: tag -> [doc_ids]
  std::unordered_map<std::string, std::vector<int>> index;

  suzume::TagOptions tag_opts;
  tag_opts.pos_filter = static_cast<std::uint8_t>(SUZUME_TAG_POS_NOUN | SUZUME_TAG_POS_VERB);
  tag_opts.exclude_basic = true;
  tag_opts.use_lemma = true;
  tag_opts.min_length = 2;

  for (const auto& doc : docs) {
    auto tags = analyzer.generateTags(doc.text, tag_opts);
    if (tags.empty()) {
      std::cerr << "tag generation returned no tags for document " << doc.id << "\n";
      return 1;
    }
    for (const auto& tag : tags) {
      index[tag.tag].push_back(doc.id);
    }
  }

  // Print index
  std::cout << "Inverted index:\n";
  for (const auto& [tag, doc_ids] : index) {
    std::cout << "  " << tag << " -> [";
    for (size_t i = 0; i < doc_ids.size(); ++i) {
      if (i > 0) {
        std::cout << ", ";
      }
      std::cout << doc_ids[i];
    }
    std::cout << "]\n";
  }

  return index.empty() ? 1 : 0;
}
